x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/mapreduce/AbstractMapReducer.java

```diff
@@ -0,0 +1,256 @@
+     * when there is only one {@linkplain InternalAggregation}.
+     */
+    boolean mustReduceOnSingleInternalAgg() {
+        return true;
+    }
+
+    /**
+     * Methods for MapReduceAggregator and InternalMapReduceAggregation.
+     *
+     * This code assumes that both execute these methods in the right order and therefore the objects are correct.
+     *
+     * (By design MapReduceAggregator and InternalMapReduceAggregation are generic and don't know the internal type)
+     */
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(aggregationName);
+    }
+
+    Writeable mapInit(BigArrays bigArrays) {
+        return doMapInit(bigArrays);
+    }
+
+    Writeable map(Stream<Tuple<String, List<Object>>> keyValues, Writeable mapReduceContext) {
+        @SuppressWarnings("unchecked")
+        MapReduceContext context = (MapReduceContext) mapReduceContext;
+
+        return doMap(keyValues, context);
+    }
+
+    Writeable mapFinalize(Writeable mapReduceContext) {
+        @SuppressWarnings("unchecked")
+        MapReduceContext context = (MapReduceContext) mapReduceContext;
+
+        return doMapFinalize(context);
+    };
+
+    Writeable reduceInit(BigArrays bigArrays) {
+        return doReduceInit(bigArrays);
+    }
+
+    Writeable reduce(Stream<Writeable> partitions, Writeable mapReduceContext) {
+        @SuppressWarnings("unchecked")
+        MapReduceContext context = (MapReduceContext) mapReduceContext;
+
+        @SuppressWarnings("unchecked")
+        Stream<MapReduceContext> p = (Stream<MapReduceContext>) partitions;
+
+        return doReduce(p, context);
+    }
+
+    Writeable combine(Stream<Writeable> partitions, Writeable mapReduceContext) {
+        @SuppressWarnings("unchecked")
+        MapReduceContext context = (MapReduceContext) mapReduceContext;
+
+        @SuppressWarnings("unchecked")
+        Stream<MapReduceContext> p = (Stream<MapReduceContext>) partitions;
+
+        return doCombine(p, context);
+    }
+
+    Writeable reduceFinalize(Writeable mapReduceContext) throws IOException {
+        @SuppressWarnings("unchecked")
+        MapReduceContext context = (MapReduceContext) mapReduceContext;
+
+        return doReduceFinalize(context);
+    };
+
+    Writeable finalizeSampling(SamplingContext samplingContext, Writeable mapReduceResult) {
+        @SuppressWarnings("unchecked")
+        MapReduceResult result = (MapReduceResult) mapReduceResult;
+
+        return doFinalizeSampling(samplingContext, result);
+    }
+
+    void writeContext(StreamOutput out, Writeable Writeable) throws IOException {
+        Writeable.writeTo(out);
+    }
+
+    MapReduceContext readMapContext(StreamInput in, BigArrays bigArrays) throws IOException {
+        // TODO: what if this isn't a map context but the result object?
+        // it seems to dangerous to assume that deserialization only happens on partial reduced objects...
```

Comment:
I think this shows one of the flaws of this design. When using the typical `InternalAgg` design for combining reduction, this concern doesn't really occur. Mainly because the aggregator itself does not allow itself to be serialized between nodes (requires that internal aggs are built first).